### PR TITLE
Identity chooser

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: components/credentials/credentials-controller.js [hunk 1]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
   var trustedDomainPrefix = 'settings.trustedDomains.';
   var trustedDomainKey = trustedDomainPrefix + self.callbackHostName;
 
+  if(
+    self.identity.id !== self.requestOwner &&
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: We treat `if` differently from functions ... we don't drop all conditions down a line like we do for all parameters, keep first condition on the same line as `if(`.

File: components/credentials/credentials-controller.js [hunk 2]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
...
+  }
+  var session = null;
+  var sessionKey = [self.identity.id.replace(/[:]/g,''), 'session'].join('.');
+  if(requestData.keyType === keyTypes.EXISTING) {
+    session = ipCookie(sessionKey);
+    if(session === undefined) {
+      // NOTE: stripping colon because browser encodes it
+      // this could be an acknowledgement where the owner is in claim.id
+      sessionKey = [self.identity.credential[0]['@graph']
+        .claim.id.replace(/[:]/g,''), 'session'].join('.');
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: This session key stuff seems either hacky or brittle. What's going on? If we can't use standard encode/decode functions, then create helper encodeKey/decodeKey functions and call those.

File: components/credentials/credentials-controller.js [hunk 3]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
...
+      session = ipCookie(sessionKey);
+    }
+    if(session === undefined) {
+      // the information in the document does not correspond with an identity
+      brAlertService.add('error', 'Active session not found.');
+      throw new Error('Error: Active session not found.');
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: 'Error: ' prefix seems unnecessary.

File: components/credentials/credentials-controller.js [hunk 4]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
...
+    }
+    // refresh session
+    ipCookie(sessionKey, session, {
+      expires: 120,
+      expirationUnit: 'minutes',
+      secure: true
+    });
+  }
+  if(requestData.keyType !== keyTypes.EXISTING) {
+    console.log('IN TEMPORARY', sessionKey);
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: Remove logging.

File: components/credentials/credentials-controller.js [hunk 5]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
...
+      if(!session) {
+        // this could be an acknowledgement where the owner is in claim.id
+        sessionKey = [self.identity.credential[0]['@graph']
+          .claim.id.replace(/[:]/g,''), 'session'].join('.');
+        session = JSON.parse(sessionStorage.getItem(sessionKey));
+      }
+      if(!session) {
+        // the information in the document does not correspond with an identity
+        brAlertService.add('error', 'Active session not found.');
+        throw new Error('Error: Active session not found.');
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: Unnecessary "Error: "

File: components/credentials/credentials-controller.js [hunk 6]
```diff
@@ -20,6 +33,58 @@ function factory($scope, config, $location, ipCookie, localStorageService) {
...
+    });
+  }
+  if(requestData.keyType !== keyTypes.EXISTING) {
+    console.log('IN TEMPORARY', sessionKey);
+    try {
+      session = JSON.parse(sessionStorage.getItem(sessionKey));
+      if(!session) {
+        // this could be an acknowledgement where the owner is in claim.id
+        sessionKey = [self.identity.credential[0]['@graph']
+          .claim.id.replace(/[:]/g,''), 'session'].join('.');
```
Comment (dlongley) [inline components/credentials/credentials-controller.js]: Use encodeKey/decodeKey or some other standard functions that would work here.

Comment (dlongley) [inline components/credentials/credentials-controller.js]: Creating a function to get a sessionKey from an identity doc would be a good idea I think. Looks like this is done multiple times and could be error prone.

File: components/request/request-controller.js [hunk 7]
```diff
@@ -175,31 +242,59 @@ function factory($scope, $http, $location, ipCookie, config, brAlertService) {
...
     var id = uuid.v4();
     var authioCallback =
       config.data.baseUri + '/credentials?id=' + id;
     var credentialCallback = $location.search().credentialCallback;
     var storageCallback = $location.search().storageCallback;
+    var credentialRequest = {};
+    var publicKeyType = ['CryptographicKey'];
+
+    if(self.keyType === keyTypes.TEMPORARY) {
+      publicKeyType.push('EmphemeralCryptographicKey');
```
Comment (dlongley) [inline components/request/request-controller.js]: Typo: EmphemeralCryptographicKey => EphemeralCryptographicKey

File: components/request/request-controller.js [hunk 8]
```diff
@@ -175,31 +242,59 @@ function factory($scope, $http, $location, ipCookie, config, brAlertService) {
...
+    }
+    if(self.keyType === keyTypes.NEW) {
+      publicKeyType.push('NewCryptographicKey');
```
Comment (dlongley) [inline components/request/request-controller.js]: Don't add this type. Just later check for presence of `id`... no `id` means key is new or ephemeral, type will indicate whether it's ephemeral or not.

File: components/request/request-controller.js [hunk 9]
```diff
@@ -175,31 +242,59 @@ function factory($scope, $http, $location, ipCookie, config, brAlertService) {
...
       }
 
-      navigator.credentials.request(config.data.credentialRequest, {
+      navigator.credentials.request(credentialRequest, {
         requestUrl: session.credentialRequestUrl,
         credentialCallback: authioCallback
       });
-    } else if(credentialCallback && config.data.sendCryptographicKeyCredential) {
+    } else if(
+      credentialCallback && config.data.sendCryptographicKeyCredential) {
```
Comment (dlongley) [inline components/request/request-controller.js]: Move up a line after the paren.

File: components/request/request.html [hunk 10]
```diff
@@ -1,36 +1,105 @@
...
-          <input type="checkbox" name="public-terminal" ng-model="model.publicComputer"/> I'm using a public computer
+<div class="row" ng-controller="RequestController as model">
+  <div ng-show="model.display.identityChooser" ng-init="model.identityChooser()">
+    <h1 class="col-md-offset-3" style="padding-left: 40px">
+      <span ng-show="!model.identitySelected">Select an Identity</span>
+      <span ng-show="model.identitySelected">Log In</span>
+    </h1>
+    <div class="col-md-offset-3 col-md-6">
+      <div ng-hide="model.identitySelected && !identity.selected"
+        class="well" ng-repeat="identity in model.identities">
```
Comment (dlongley) [inline components/request/request.html]: Does the identity selector look better if you remove `class="well"` either here or below?

File: components/request/request.html [hunk 11]
```diff
@@ -1,36 +1,105 @@
...
+        <div ng-show="identity.selected"
+          style="padding-bottom: 5px; border-color: #aaaaaa; border-bottom-style: solid; border-bottom-width: 1px; margin-bottom: 15px;">
+          <span ng-click="model.showAll(identity, loginForm);" style="cursor: pointer;">
+            <i class="fa fa-reply-all"></i><span style="padding-left: 5px">show all</span>
+          </span>
         </div>
-
-        <div style="text-align:center;">
-          <button ng-disabled="model.generating" ng-click="model.login(model.username,model.password)" type="submit" class="btn btn-primary">Login</button>
+        <div class="well" ng-class="{'br-selectable': !model.identitySelected}" ng-click="model.redirect(identity, loginForm)" style="cursor: pointer;">
```
Comment (dlongley) [inline components/request/request.html]: Does the identity selector look better if you remove `class="well"` either here or above?

File: components/request/request.html [hunk 12]
```diff
@@ -1,36 +1,105 @@
...
+          <img src="http://simpleicon.com/wp-content/uploads/favorite-user-64x64.png">
```
Comment (dlongley) [inline components/request/request.html]: Use font-awesome instead: http://fortawesome.github.io/Font-Awesome/3.2.1/cheatsheet/

Probably something like `icon-user`.

File: components/request/request.html [hunk 13]
```diff
@@ -1,36 +1,105 @@
...
+          <span style="margin-left: 30px"><strong>{{ identity.label }}</strong></span>
```
Comment (dlongley) [inline components/request/request.html]: By using bootstrap and font awesome, you probably don't have to worry about doing inline CSS style stuff like setting margins. Just use font awesome and choose the appropriate bootstrapism instead for combining an icon with something.

File: configs/authorization.io.js [hunk 14]
```diff
@@ -89,4 +89,23 @@ config.requirejs.bower.packages.push({
...
+  config.views.vars['bedrock-angular-credential'].libraries || {};
+config.views.vars['bedrock-angular-credential'].libraries.default = {
+  vocabs: [
+    config.server.baseUri + '/vocabs/test-v1.jsonld'
+  ]
+};
+
+// lower minimum wait time for proofs for demo
+config.authio.proofs.minWaitTimeInSecs = 2;
+config.authio.proofs.maxWaitTimeInSecs = 3;
```
Comment (dlongley) [inline configs/authorization.io.js]: Why are all these changes in here? How are they related to the identity chooser?

**Comment by mattcollier** [comment]
@dlongley updated based on your comments.  I removed the use of 'newCryptographicKey'.  You referred to some 'id' not being present in this case, I'm not sure what you're referring to, so additional modification may be required.

**Pull request closed without merging**